### ISSUE TITLE
Appreciation thread :D

### ISSUE THREAD

**Issue by otse** [schteppe/cannon.js#295]
Not trynna abuse issues here, but this looks too good to be true. Out of all the millions of broken box2d ports (which is a nightmare, I know it's 2d its in the name), this looks extremely promising. (I created gta2.net with some box2d port and it was the most awful thing ever)

Can you keep this open real quick, I might have some questions

I'm applying it to http://narrow.ak.fm/
I got here from http://stackoverflow.com/questions/13960577/how-to-let-the-player-control-collide-with-walls-in-three-js

**Comment by otse** [comment]
How do I go about making stairs? The gravity of -20 will roll the sphere right off.

Do I have to hack things?

I tried changing the sphereBody in the fps example into a box, but the box would nearly refuse to move.

Some ideas would be grateful.

**Comment by schteppe** [comment]
What are you trying to make? A character controller that can climb stairs?
Usually, using a physics engine for more complicated character movement is not a good idea. This article explains a few related problems: http://www.learn-cocos2d.com/2013/08/physics-engine-platformer-terrible-idea/

In Bullet physics there's a btKinematicCharacterController class, that solves these problems. but it's probably not possible to implement that in cannon.js (need convex sweep collision detection).

**Comment by schteppe** [comment]
Oh and if you only want the box to glide on the floor, just adjust the friction coefficient.

**Comment by otse** [comment]
Nice, will try tomorrow. Thanks for the great answer!

Op vrijdag 8 juli 2016 heeft Stefan Hedman notifications@github.com het
volgende geschreven:

> Oh and if you only want the box to glide on the floor, just adjust the
> friction coefficient.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/schteppe/cannon.js/issues/295#issuecomment-231461149,
> or mute the thread
> https://github.com/notifications/unsubscribe/ABEkIslaw3SkQpVPusfH1qAQxkBScCymks5qTq-qgaJpZM4JH8qb
> .

## 

Sent from Gmail Mobile

**Comment by otse** [comment]
Thanks, yeah, that ought to make a moving box (à la HL2).

Maybe check out this version: http://narrow.ak.fm#showPhysics, this gives static level cannon.js-bodies a texture.

I'd like to use Cannon at least for Rollermine-type enemies. Is  there a way I can run an isolated set of routines, exempting a single, possibly static `playerBody` from the normal rules? Is there a way to hack your library like this?

E.g. if my level design has no drops near stairs, I can apply maybe some sort of pulley at the base of the stairs.

**Comment by otse** [comment]
Changing the friction coefficient to this

``` Javascript
physicsContactMaterial = new CANNON.ContactMaterial(physicsMaterial, physicsMaterial,
            0.5, // friction coefficient
            0.0  // restitution
)
```

Gives me a peculiar js error:
`cannon.min.js:27 Uncaught TypeError: Cannot use 'in' operator to search for 'friction' in 0.5.`

Is the fps example outdated? *\* Oh, the third parameter is supposed to be an object.

**Comment by otse** [comment]
Or is it possible to always keep the `playerBody` box upright? Granted that it's rectangular, and I don't want it to topple over. Currently I do this, which causes hopping

``` Javascript
nr.cannon.plybody.quaternion.setFromAxisAngle(new CANNON.Vec3(0,1,0), 0)
```

When it touches a stairs, maybe simply teleport it up? Make each step rectangular. Keep rotation fixed.

**Comment by schteppe** [comment]
To disable rotation you can use the fixedRotation feature. See the demo: https://github.com/schteppe/cannon.js/blob/master/demos/fixedRotation.html

**Comment by otse** [comment]
Oh, there is an examples folder _and_ a demos folder

**Comment by otse** [comment]
Okay, the fixedRotation box is way nicer than the orb, but the player now refuses to go uphill :dancer:

**Comment by otse** [comment]
I fixed the stairs with the following callback. Also, walking down the stairs does not give trippy behavior. This is the most straightforward idea I had and went with it, and it works 1:1.

Walking up stairs works perfectly now http://narrow.ak.fm/#showPhysics

Here is the callback I used

``` CoffeeScript
if 'Stair' is e.mat
    body.addEventListener 'collide', (e) ->
        that = this

        return if @nr.skip

        @nr.skip = true

        setTimeout ->
            that.nr.skip = no
        , 250

        plybody = nr.cannon.plybody

        plybody.position.y = @position.y+.47

        return
```

To refrain bumpy behavior, I use the timeout to avoid an immediate second/third invocation of the collider event, this prevents ADHD.

**Comment by RaeveNoir** [comment]
A pill shape (cylinder with hemispheres on top and bottom) is the standard FPS collision shape. It will solve most of your problems.

**Comment by schteppe** [comment]
I believe you're right @Ranzear, a pill/capsule shape would solve many problems. Too bad it's not implemented in Cannon.js yet... However, just using a few spheres in a compound shape should be fine too.

**Comment by otse** [comment]
Would that resolve the snappy staircase?

Op maandag 11 juli 2016 heeft Stefan Hedman notifications@github.com het
volgende geschreven:

> I believe you're right @Ranzear https://github.com/Ranzear, a
> pill/capsule shape would solve many problems. Too bad it's not implemented
> in Cannon.js yet... However, just using a few spheres in a compound shape
> should be fine too.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/schteppe/cannon.js/issues/295#issuecomment-231711313,
> or mute the thread
> https://github.com/notifications/unsubscribe/ABEkIpaBU0gh7L_-qu_WBpB0Y26pfoFmks5qUiv8gaJpZM4JH8qb
> .

## 

Sent from Gmail Mobile

**Comment by RaeveNoir** [comment]
Yes. The radius of the bottom of the pill will affect how smoothly and easily it goes up and down steps and slopes. The round top prevents other actors from standing on top of it. You can approximate this with two spheres and not likely need the cylinder joining them, and even use a 'snowman' (legs, body, head) which doubles as simplified hit detection for weaponry and physics damage. I like to ignore limbs, so only the body and head calculate damage.

**Comment by otse** [comment]
That sounds very good. I'd need some new code for the steps then though

You guys like the new GitHub font lol?
